### PR TITLE
OpamFile: Make all writes atomic

### PR DESCRIPTION
**Pull request by kit-ty-kate**
A continuation of https://github.com/ocaml/opam/pull/5476#pullrequestreview-1351056585

Partially reverts 64a569feee43c3870c84558db291abf780a70bf7

### CODE REVIEW

**Comment by rjbou** [comment]
Do we know if there is any performance regression?

**Comment by kit-ty-kate** [comment]
> Do we know if there is any performance regression?

All these files are written (in the worse case scenario) a couple times per opam calls and `rename` is a fairly fast system call, so no.

The actual worse case scenario is for `opam admin add-constraint` if you give something like `ocaml<3` but even for that, i've just tried it and it's actually faster (2.9s vs 3.7s before)

**Review by rjbou — APPROVED** [review]
On the code and idea, lgtm!
The commit and PR comment should mention that it is a partial revert of https://github.com/ocaml/opam/pull/5476/commits/64a569feee43c3870c84558db291abf780a70bf7, and with deprecation notice on `OpamFilename.with_open_out_bin`, it is good to merge!

File: src/core/opamFilename.mli [hunk 1]
```diff
@@ -143,13 +143,9 @@ val open_in_bin: t -> in_channel
 val open_out: t -> out_channel
 val open_out_bin: t -> out_channel
 
-(** [with_open_out_bin filename f] opens [f] and passes the out_channel to [f].
-    If [f] raises an exception, then [filename] is deleted and then exception is
+(** [with_open_out_bin_atomic filename f] opens [f] and passes the out_channel to [f].
+    If [f] raises an exception, then [filename] will be unaltered and then exception is
     propagated. The out_channel does not have to be closed by [f]. *)
-val with_open_out_bin: t -> (out_channel -> unit) -> unit
```
[the remark this answers is not in the document]
Comment (kit-ty-kate) [reply src/core/opamFilename.mli:149]: is that really necessary given this function was only added in opam 2.2?

[the remark this answers is not in the document]
Comment (rjbou) [reply src/core/opamFilename.mli:149]: It is released, not only in pre-release. We can also just keep it and not use it for OpamFile reading

[the remark this answers is not in the document]
Comment (kit-ty-kate) [reply src/core/opamFilename.mli:149]: done

**Pull request merged**